middle_end/flambda2.0/from_lambda/lambda_to_flambda_primitives.ml

```diff
@@ -808,21 +883,52 @@ let convert_lprim ~backend (prim : L.primitive) (args : Simple.t list)
         Prim (Unary (Unbox_number Naked_nativeint, arg)))))
   | Pint_as_pointer, [arg] -> Unary (Int_as_pointer, arg)
   | Pbigarrayref (unsafe, num_dimensions, kind, layout), args ->
-    (* CR mshinwell: When num_dimensions = 1 then we could actually
-       put the bounds check in Flambda. *)
-    let is_safe : P.is_safe = if unsafe then Unsafe else Safe in
-    let kind = C.convert_bigarray_kind kind in
-    let layout = C.convert_bigarray_layout layout in
-    let box = bigarray_box_raw_value_read kind in
-    box (Variadic (Bigarray_load (is_safe, num_dimensions, kind, layout), args))
+    begin match C.convert_bigarray_kind kind,
+                C.convert_bigarray_layout layout with
+    | Some kind, Some layout ->
+      let b, indexes =
+        match args with
+        | b :: indexes ->
+          if num_dimensions <> List.length indexes then
+            Misc.fatal_errorf "Bad typing for Pbigarrayref";
+          b, indexes
+        | [] -> Misc.fatal_errorf "Wrong arity for Pbigarrayref"
+      in
+      let box = bigarray_box_raw_value_read kind in
+      box (bigarray_ref ~dbg ~unsafe kind layout b indexes)
+    | None, _ ->
+      Misc.fatal_errorf
+        "Lambda_to_flambda_primitives.convert_lprim: Pbigarrayref primitives \
+         with an unknown kind should have been removed by Prepare_lambda."
+    | _, None ->
+      Misc.fatal_errorf
+        "Lambda_to_flambda_primitives.convert_lprim: Pbigarrayref primitives \
+         with an unknown layout should have been removed by Prepare_lambda."
+    end
   | Pbigarrayset (unsafe, num_dimensions, kind, layout), args ->
-    let is_safe : P.is_safe = if unsafe then Unsafe else Safe in
-    let kind = C.convert_bigarray_kind kind in
-    let layout = C.convert_bigarray_layout layout in
-    let unbox = bigarray_unbox_value_to_store kind in
-    let indexes, value_to_store = Misc.split_last args in
-    Variadic (Bigarray_set (is_safe, num_dimensions, kind, layout),
-              indexes @ [unbox value_to_store])
+    begin match C.convert_bigarray_kind kind,
+                C.convert_bigarray_layout layout with
+    | Some kind, Some layout ->
+      let b, indexes, value =
+        match args with
+        | b :: args ->
+          let indexes, value = Misc.split_last args in
+          if num_dimensions <> List.length indexes then
```

Comment:
FYI There is a new(ish) function `List.compare_length_with` for this, which can be more efficient.